server/dataset_context.cc

```diff
@@ -133,6 +133,130 @@ doGetColumn(const Utf8String & tableName,
 /* ROW EXPRESSION DATASET CONTEXT                                            */
 /*****************************************************************************/
 
+RowName
+SqlExpressionDatasetScope::RowScope::
+getRowName() const
+{
+    if (rowName) return *rowName;
+    else {
+        ExcAssert(row);
+        return row->rowName;
+    }
+}
+
+const RowName &
+SqlExpressionDatasetScope::RowScope::
+getRowName(RowName & storage) const
+{
+    if (rowName) return *rowName;
+    else return row->rowName;
+}
+
+RowHash
+SqlExpressionDatasetScope::RowScope::
+getRowHash() const
+{
+    if (rowName) return *rowName;
+    else return row->rowHash;
+}
+
+const ExpressionValue &
+SqlExpressionDatasetScope::RowScope::
+getColumn(const ColumnName & columnName,
+          const VariableFilter & filter,
+          ExpressionValue & storage,
+          ssize_t knownOffset) const
+{
+    if (expr) {
+        const ExpressionValue * val
+            = expr->tryGetNestedColumn(columnName, storage, filter);
+        if (!val)
+            return storage = ExpressionValue::null(Date::negativeInfinity());
+        else return *val;
+    }
+    else {
+        const ExpressionValue * fromOutput
+            = searchRow(row->columns, columnName, filter, storage);
+        if (fromOutput)
+            return *fromOutput;
+        
+        return storage = std::move(ExpressionValue::null(Date::negativeInfinity()));
+    }
+}
+
+ExpressionValue
+SqlExpressionDatasetScope::RowScope::
+getColumnCount() const
+{
+    if (row) {
+        ML::Lightweight_Hash_Set<ColumnHash> columns;
+        Date ts = Date::negativeInfinity();
+        
+        for (auto & c: row->columns) {
+            columns.insert(std::get<0>(c));
+            ts.setMax(std::get<2>(c));
+        }
+
+        return ExpressionValue(columns.size(), ts);
+    }
+    else {
+        throw HttpReturnException(600, "getColumnCount() from ExpressionValue");
+    }
+}
+        
+ExpressionValue
+SqlExpressionDatasetScope::RowScope::
+getFilteredValue(const VariableFilter & filter) const
+{
+    if (row) {
+        // TODO: if one day we are able to prove that this is
+        // the only expression that touches the row, we could
+        // move it into place
+        ExpressionValue val(row->columns);
+        return val.getFilteredDestructive(filter);
```

Comment:
We would need much stronger static analysis in MLDB to prove that something is moveable.  For the moment, it's const and we shouldn't move.